cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTableMigrator.java

```diff
@@ -0,0 +1,175 @@
+    HTable newIndexTable = tableUtil.createHTable(conf, indexTableId);
+    HTable newDataTable = tableUtil.createHTable(conf, dataTableId);
+
+    LOG.info("Starting upgrade for table {}", Bytes.toString(oldTable.getTableName()));
+    try {
+      ScanBuilder scan = tableUtil.buildScan();
+      scan.setTimeRange(0, HConstants.LATEST_TIMESTAMP);
+      scan.setMaxVersions(1); // we only need to see one version of each row
+      try (ResultScanner resultScanner = oldTable.getScanner(scan.build())) {
+        Result result;
+        while ((result = resultScanner.next()) != null) {
+          Put put = new Put(result.getRow());
+
+
+          Long hBaseVersion = null;
+
+          for (Map.Entry<byte[], NavigableMap<byte[], NavigableMap<Long, byte[]>>> familyMap :
+            result.getMap().entrySet()) {
+            for (Map.Entry<byte[], NavigableMap<Long, byte[]>> columnMap : familyMap.getValue().entrySet()) {
+              for (Map.Entry<Long, byte[]> columnEntry : columnMap.getValue().entrySet()) {
+                Long timeStamp = columnEntry.getKey();
+                byte[] colVal = columnEntry.getValue();
+                put.add(familyMap.getKey(), columnMap.getKey(), timeStamp, colVal);
+                LOG.info("Migrating row. family: '{}', column: '{}', ts: '{}', colVal: '{}'",
+                         familyMap.getKey(), columnMap.getKey(), timeStamp, colVal);
+
+                // the hbase version for all the columns are equal since every column of a row corresponds to one
+                // partition and a partition is created entirely within a single transaction.
+                hBaseVersion = timeStamp;
+              }
+            }
+          }
+
+          Preconditions.checkNotNull(hBaseVersion,
+                                     "There should have been at least one column in the scan. Table: {}, Rowkey: {}",
+                                     oldPartitionsTableId, result.getRow());
+
+          // additionally since 3.1.0, we keep two addition columns for each partition: creation time of the partition
+          // and the transaction write pointer of the transaction in which the partition was added
+          byte[] hbaseVersionBytes = Bytes.toBytes(hBaseVersion);
+          put.add(columnFamily, PartitionedFileSetDataset.WRITE_PTR_COL,
+                  hBaseVersion, hbaseVersionBytes);
+          // here, we make the assumption that dropping the six right-most digits of the transaction write pointer
+          // yields the timestamp at which it started
+          byte[] creationTimeBytes = Bytes.toBytes(hBaseVersion / MILLION);
+          put.add(columnFamily, PartitionedFileSetDataset.CREATION_TIME_COL,
+                  hBaseVersion, creationTimeBytes);
+
+          newDataTable.put(put);
+
+          // index the data table on the two columns
+          Put indexTableWritePointer = new Put(IndexedTable.createIndexKey(result.getRow(),
```

Comment:
call this variable indexPut instead of indexTableWritePointer? it is a put, not a write pointer...